connection.go

```diff
@@ -129,6 +163,151 @@ func (c *conn) ExecContext(ctx context.Context, query string, args []driver.Name
 }
 
+func Succeeded(response *http.Response) bool {
+	if response.StatusCode == 200 || response.StatusCode == 201 || response.StatusCode == 202 || response.StatusCode == 204 {
+		return true
+	}
+	return false
+}
+
+func (c *conn) HandleStagingPut(ctx context.Context, presignedUrl string, headers map[string]string, localFile string) (driver.Result, error) {
+	if localFile == "" {
+		return nil, dbsqlerrint.NewDriverError(ctx, "cannot perform PUT without specifying a local_file", nil)
+	}
+	client := &http.Client{}
+
+	dat, err := os.ReadFile(localFile)
+
+	req, _ := http.NewRequest("PUT", presignedUrl, bytes.NewReader(dat))
+
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	content, err := io.ReadAll(res.Body)
+
+	if err != nil || !Succeeded(res) {
+		return nil, dbsqlerrint.NewDriverError(ctx, fmt.Sprintf("staging operation over HTTP was unsuccessful: %d-%s, nil", res.StatusCode, content), nil)
+	}
+	return driver.ResultNoRows, nil
+
+}
+
+func (c *conn) HandleStagingGet(ctx context.Context, presignedUrl string, headers map[string]string, localFile string) (driver.Result, error) {
+	if localFile == "" {
+		return nil, fmt.Errorf("cannot perform GET without specifying a local_file")
+	}
+	client := &http.Client{}
+	req, _ := http.NewRequest("GET", presignedUrl, nil)
+
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	content, err := io.ReadAll(res.Body)
+
+	if err != nil || !Succeeded(res) {
+		return nil, dbsqlerrint.NewDriverError(ctx, fmt.Sprintf("staging operation over HTTP was unsuccessful: %d-%s, nil", res.StatusCode, content), nil)
+	}
+
+	err = os.WriteFile(localFile, content, 0644) //nolint:gosec
+	if err != nil {
+		return nil, err
+	}
+	return driver.ResultNoRows, nil
+
+}
+
+func (c *conn) HandleStagingDelete(ctx context.Context, presignedUrl string, headers map[string]string) (driver.Result, error) {
+	client := &http.Client{}
+	req, _ := http.NewRequest("DELETE", presignedUrl, nil)
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	content, err := io.ReadAll(res.Body)
+
+	if err != nil || !Succeeded(res) {
+		return nil, dbsqlerrint.NewDriverError(ctx, fmt.Sprintf("staging operation over HTTP was unsuccessful: %d-%s, nil", res.StatusCode, content), nil)
+	}
+
+	return driver.ResultNoRows, nil
+}
+
+func localPathIsAllowed(ctx context.Context, localFile string) bool {
+	stagingAllowedLocalPaths := driverctx.StagingPathsFromContext(ctx)
+	for i := range stagingAllowedLocalPaths {
+		path := stagingAllowedLocalPaths[i]
+		relativePath, err := filepath.Rel(path, localFile)
+		if err != nil {
+			return false
+		}
+		if !strings.Contains(relativePath, "../") {
+			return true
+		}
```

Comment:
Do you have unit or e2e tests that prove this is secure?

I'm thinking something along these lines: https://github.com/databricks/databricks-sql-python/blob/fcc262f9bd4f48432e599b4f4f8b9cf62f8ace73/tests/e2e/common/staging_ingestion_tests.py#L213-L228

Rather than manually checking for `../` I'd recommend using a method like Python's `os.path.commonpath` which verifies absolutely that the allowed local path isn't being gamed. It does this by actually resolving the `../` and comparing the output absolute path to the allowed local paths.

From what I'm seeing [here](https://stackoverflow.com/questions/28024731/check-if-given-path-is-a-subdirectory-of-another-in-golang) I think a little more is required beyond just using go's `Rel` method.

Reply:
I can add in that unit test. Unfortunately, I'm not sure what this analog would be. The article you linked says that we just need to look for the ../ right?